src/detectors/Rust/RustComponentDetector.ts

```diff
@@ -0,0 +1,71 @@
+import { IProjectComponentDetector } from '../IProjectComponentDetector';
+import { injectable, inject } from 'inversify';
+import { Types } from '../../types';
+import { LanguageAtPath, ProjectComponent, ProjectComponentFramework, ProjectComponentPlatform, ProjectComponentType } from '../../model';
+import { IFileInspector, RustPackageInspector } from '../../inspectors';
+import lo from 'lodash';
+
+@injectable()
+export class RustComponentDetector implements IProjectComponentDetector {
+  constructor(
+    private readonly packageInspector: RustPackageInspector,
+    @inject(Types.IFileInspector) private readonly fileInspector: IFileInspector,
+  ) {}
+
+  async detectComponent(langAtPath: LanguageAtPath): Promise<ProjectComponent[]> {
+    const result: ProjectComponent[] = [];
+
+    // Cargo outputs a binary output when it finds:
+    // * `src/main.rs`
+    // * `src/bin/*.rs`
+    // * `src/bin/*/main.rs`
+    // * a `bin` entry in the manifest
+    const hasBinaryFile = Promise.all([
+      this.fileInspector.exists('src/main.rs'),
+      this.fileInspector
+        .readDirectory('src/bin')
+        .then(async (entries) => {
```

Comment:
I honestly believe that the functional style programming pattern here is more readable, as it better demonstrates the flow of information. With try/catch I'd either have to include too much code into the block or use a `let` binding which we rust people hate (joke).

As for the `Promise.all` call, I would really rather want some kind of a `Promise.raceWhile` to short-circuit on the first rejection or the first resolution with the value of `true`.

Anyhow, the alternative code would look like this:
```
const hasMain = await this.fileInspector.exists('src/main.rs');
let hasSrcBin = false;
try {
    const binFolderEntries = await this.fileInspector.readDirectory('src/bin');
    for (const entry of binFolderEntries) {
      const entryPath = `src/bin/${entry}`;
      if (
        (entry.endsWith('.rs') && (await this.fileInspector.isFile(entryPath))) ||
        (await this.fileInspector.isFile(`${entryPath}/main.rs`))
      ) {
        hasSrcBin = true;
      }
    }
} catch (e) {
    if (e.code === 'ENOENT') {
        // ignore
    } else {
        throw e;
    }
}
const hasManifestBin = _.get(this.packageInspector.cargoManifest, 'bin.length', 0) > 0;
const hasBinaryFile = hasMain || hasSrcBin || hasManifestBin;
```

I personally don't like the monstrous try/catch and the fact that adding another condition would mean having to add the condition calculation and then updating the `hasBinaryFile` which is now all handled by the `Promise.all([]).then(a => a.some(<identity>)` function and I don't even have to come up with additional variable names :D

Either way, if you insist on strictly using imperative `async/await` with `try/catch` I'll do the second variant.

Reply:
Hi👋 Nice job 👍 Please stick to the `async/await`. Main reason is to avoid `Promise.all` (we already have some problems with it). Also you probably can break out of the `for` loop after you set `hasSrc = true`. Or you can use `array.some()`

Reply:
Okay, I'll change it. Just out of curiosity, may I know what your problems with `Promise.all` are?

I can't really use `array.some()` because I do `await` inside the check loop but the break is a fair point. 👍

Reply:
The problem was, that when we used the `Promise.all()` in the other project with more async methods inside (each returns a different return type) we had to unforce the types because it was translated to `undefined` (see [this issue](https://github.com/microsoft/TypeScript/issues/33752)). When we add more than 10 methods inside the `Promise.all()` it crashed. There was an unreasonable cap number of methods we can pass inside the `Promise.all()`.
To sum it up: 
1) There was a problem with types.
2) There was a problem with the number of passed methods in `Promise.all()`.
Maybe there were more, but I don't remember it. Maybe @Jakub-Vacek ?

Reply:
Hmm I see, It is defined in the typescript library with overloads for up to `T10`
```ts
all<T1, T2, T3, T4, T5, T6, T7, T8, T9, T10>(values: readonly [T1 | PromiseLike<T1>, T2 | PromiseLike<T2>, T3 | PromiseLike<T3>, T4 | PromiseLike<T4>, T5 | PromiseLike<T5>, T6 | PromiseLike<T6>, T7 | PromiseLike<T7>, T8 | PromiseLike<T8>, T9 | PromiseLike<T9>, T10 | PromiseLike<T10>]): Promise<[T1, T2, T3, T4, T5, T6, T7, T8, T9, T10]>;
```
In my opinion it is not unreasonable, it is simply arbitrary because there are no variadic generics in typescript (yet! But we did get variadic tuple generics in TS4). For example, Rust does the same thing for tuples up to some limit because it also doesn't have variadic generics. I think it's better than nothing.

However I'd also like to refute this: This only applies to when you pass a tuple of more than 10 promises and at least two of them have different return types. Then instead of one of the typed overloads you'll invoke the `all<T>(values: readonly (T | PromiseLike<T>)[]): Promise<T[]>;` overload which requires that each promise has type `T` so you'll have to define `T` as an union of all the types appearing in the tuple. This is of course a problem when you rely on typing like this, for example in the patterns like `const [foo, bar] = Promise.all([fooPromise, barPromise]);`.

However, my solution does not. All promises passed into the `Promise.all` call return `boolean`, which is actually also apparent by the subsequent call to `.then((a) => a.some((i) => i))` (identity function => checking for truthyness). I'll of course still respect your preference with `async/await` but I wanted to make sure it is clear that this is not the same case as you ran into :)